subprojects/plugin-development/src/test/groovy/org/gradle/plugin/devel/internal/precompiled/PreCompiledScriptTest.groovy

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.plugin.devel.internal.precompiled
+
+import org.gradle.plugin.internal.InvalidPluginIdException
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class PreCompiledScriptTest extends Specification {
+
+    def "throws for filename yielding invalid plugin id"() {
+        when:
+        new PreCompiledScript(new File("/foo/bar/f%izzbuzz.gradle"))
+
+        then:
+        thrown InvalidPluginIdException
+    }
+
+    @Unroll
+    def "creates valid java classname '#javaClass' from script filename based plugin id: #filename"() {
+        expect:
+        def script = new PreCompiledScript(new File("/foo/bar/$filename"))
+        script.generatedPluginClassName == javaClass
+
+        where:
+        filename                     | javaClass
+        "foo.gradle"                 | "FooPlugin"
+        "foo.bar.gradle"             | "BarPlugin"
```

Comment:
❓ wouldn't `foo.bar.gradle` conflicts with `baz.bar.gradle` both generating a `BarPlugin` class?

Reply:
Of course they will conflict. I must have been thinking about putting the generated classes in packages based on the pluginId namespace. Fixed now, thanks!